Add Co2signal sensor

danielsjf opened the pull request:
## Description:
This PR adds the CO2 signal component. The sensor reports the CO2 intensity (in CO2eq/kWh) in a specific region based upon the actual generation in that region and the imports and exports with neighboring regions. For instance, the sensor could be used to start devices if the CO2 emissions go below a certain threshold (e.g. charging an electrical vehicle).

The data can be checked on https://www.electricitymap.org/.

A token for the API is available via https://www.co2signal.com/.


**Related issue (if applicable):**
Not applicable

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#8037

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: co2signal
    token: SECRET # get yours for free (for personal use) from co2signal.org
    country_code: BE
    refresh_rate: 15 # minutes between polling
```

## Checklist:
  - [x] The code change is tested and works locally (as custom component).
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

danielsjf committed:
Merge pull request #1 from home-assistant/dev

Merge

danielsjf committed:
Initial commit for the co2signal sensor

danielsjf committed:
Clean code

danielsjf committed:
Run script gen_requirements_all.py

danielsjf committed:
remove unintended character

danielsjf committed:
Remove redundancy

danielsjf committed:
Remove unused imports

danielsjf committed:
Code style

danielsjf committed:
Code style fixes

homeassistant commented:
Hi @danielsjf, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#19204).

Once you do that we will be able to review and accept this pull request.

Thanks!

danielsjf committed:
Code style

homeassistant commented:
Hi @danielsjf, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#19204).

Once you do that we will be able to review and accept this pull request.

Thanks!

fabaff commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,133 @@
+For more details about this platform, please refer to the documentation
+"""
+import logging
+import voluptuous as vol
```
Import should be grouped.

MartinHjelmare replied:
Please add a blank line between standard library, 3rd party and homeassistant import groups.

fabaff commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,133 @@
+    else:
+        lat = config.get(CONF_LATITUDE)
+        lon = config.get(CONF_LONGITUDE)
+        location_type = 'coordinates'
```
Looks like that this could be handled by the validation (`vol.Inclusive` and `vol.Exclusive`).

danielsjf replied:
I've updated this part of the code. Is this what you meant?

danielsjf replied:
@fabaff I've only used vol.Inclusive as I couldn't parse the default scenario with the HASS coordinates otherwise. Currently it works with all three setups: no coordinates, coordinates and country codes.

fabaff commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,133 @@
+    @property
+    def device_class(self):
+        """Return the device class."""
+        """Icon to use in the frontend, if any."""
```
Remove this docstring.

fabaff commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,133 @@
+    def device_class(self):
+        """Return the device class."""
+        """Icon to use in the frontend, if any."""
+        return 'carbon_dioxide_intensity'
```
This is not a valid `device_class` -> https://www.home-assistant.io/components/binary_sensor/#device-class

fabaff commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,133 @@
+
+        _LOGGER.debug("Initialise %s", self._friendly_name)
+
+        self.update()
```
Remove this and add `True` as second parameter to `add_entities()` -> https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#5-entity

fabaff commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,133 @@
+                          location_type,
+                          refresh_rate))
+
+    add_devices(devs)
```
This was renamed to `add_entities` across all platforms. Please use `add_entities`.

fabaff commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,133 @@
+        if country_code is not None:
+            self._device_name = country_code
+        self._friendly_name = 'CO2 intensity - {}'.format(self._device_name)
+        self._refresh_rate = refresh_rate
```
This is not used.

fabaff commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,133 @@
+    vol.Optional(CONF_COUNTRY_CODE, default=None): cv.string,
+    vol.Optional(CONF_LATITUDE, default=0): cv.latitude,
+    vol.Optional(CONF_LONGITUDE, default=0): cv.longitude,
+    vol.Optional(CONF_REFRESH, default=60): cv.positive_int,
```
Can already be handled by `SCAN_INTERVAL`.

frenck commented:
Could not find any related documentation PR in our documentation repository. Adding `docs-missing` label.

danielsjf committed:
Fix comments PR

Comments by @fabaff
* Remove redundant comments and variables
* Follow the latest home-assistant guidelines

danielsjf committed:
Bump CO2Signal version

danielsjf committed:
Round API result

danielsjf committed:
Improve default latitude/longitude handling

danielsjf committed:
Improve friendly name

danielsjf committed:
Improve config handling

danielsjf committed:
Make lines shorter

danielsjf committed:
Style

danielsjf committed:
Convert default to variable

None does not pass cv.string

danielsjf committed:
Message if not inclusive

danielsjf committed:
Shorten line

danielsjf commented:
@fabaff this one is ready for another review.

danielsjf committed:
Update requirements

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+
+from homeassistant.const import (
+    CONF_TOKEN, CONF_LATITUDE, CONF_LONGITUDE)
+from homeassistant.helpers.config_validation import PLATFORM_SCHEMA
```
Import the platform schema from the sensor base component.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+
+ATTRIBUTION = 'Data provided by CO2signal'
+
+msg_location = "Please use either coordinates or the country code. " \
```
This variable should be a constant in all capital letter.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+    vol.Required(CONF_TOKEN): cv.string,
+    vol.Inclusive(CONF_LATITUDE, 'coords', msg=msg_location): cv.latitude,
+    vol.Inclusive(CONF_LONGITUDE, 'coords', msg=msg_location): cv.longitude,
+    vol.Optional(CONF_COUNTRY_CODE, default='None'): cv.string,
```
Remove default.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the CO2signal sensor."""
+
+    _LOGGER.debug("Setting up the CO2signal platform")
```
Remove this. It's already logged by the core.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+
+    _LOGGER.debug("Setting up the CO2signal platform")
+
+    token = config.get(CONF_TOKEN)
```
Use `dict[key]` for required config keys and keys with default config schema values.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+    lon = config.get(CONF_LONGITUDE, hass.config.longitude)
+    country_code = config.get(CONF_COUNTRY_CODE)
+
+    if country_code != 'None':
```
Check for `is not None`.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+    def __init__(self, token, country_code, lat, lon,
+                 location_type):
+        """Initialize the sensor."""
+
```
Remove blank line.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+                          lat,
+                          lon,
+                          location_type))
+
```
Why all the blank lines here?

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+    @property
+    def state(self):
+        """Return the state of the device."""
+        _LOGGER.debug("Get state for %s", self._friendly_name)
```
No logging in state properties.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+                                             latitude=self._latitude,
+                                             longitude=self._longitude)
+        else:
+            raise ValueError("Unknown location type: {location_type}"
```
When could this happen?

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+    @property
+    def device_class(self):
+        """Return the device class."""
+        return None
```
If we don't use this, we should remove it.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+
+        self._friendly_name = 'CO2 intensity - {}'.format(self._device_name)
+
+        _LOGGER.debug("Initialise %s", self._friendly_name)
```
Remove logging. We don't want side effects in init method.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+        self._latitude = lat
+        self._longitude = lon
+        self._location_type = location_type
+        self._unit = 'CO2eq/kWh'
```
This could be extracted to a constant at the module level. Then we don't need the instance attribute but can use the constant directly.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,131 @@
+    country_code = config.get(CONF_COUNTRY_CODE)
+
+    if country_code != 'None':
+        location_type = 'country_code'
```
Instead of using this variable, we can just check if `country_code` is `None`. The `location_type` variable isn't saving us any code and it's not directly used.

danielsjf committed:
Update co2signal.py

Group imports;  remove empty lines; refactor use of location_type; remove logging messages; remove unused functions; add global variables

danielsjf committed:
Update co2signal.py

Import platform schema from sensor

danielsjf committed:
Small fix

danielsjf committed:
Update co2signal.py

Remove last mentions of location_type

danielsjf commented:
@MartinHjelmare thanks for reviewing this PR. I've made the requested modifications.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,109 @@
+
+        if self._country_code is not None:
+            self._data = CO2Signal\
+                .get_latest_carbon_intensity(self._token,
```
Please break the line after the first parenthesis instead.
```py
self._data = CO2Signal.get_latest_carbon_intensity(
    self._token, country_code=self._country_code)
```

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,109 @@
+                                             country_code=self._country_code)
+        else:
+            self._data = CO2Signal\
+                .get_latest_carbon_intensity(self._token,
```
See above.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,109 @@
+import voluptuous as vol
+
+import homeassistant.helpers.config_validation as cv
+
```
Remove this blank line.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,109 @@
+        self._longitude = lon
+
+        if country_code is not None:
+            self._device_name = country_code
```
`_device_name` is only used locally here, so it doesn't need to be an instance attribute. Just make it a local variable.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,109 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTRIBUTION = 'Data provided by CO2signal'
```
This isn't used. It should be used in the `device_state_attributes` property.

danielsjf committed:
Review changes

Add attribution
Formatting

danielsjf commented:
@MartinHjelmare Thanks again for the review. I've taken the suggestions into account. Could you have another look? Especially the attribution was a bit vague to me. Different components seem to handle seemingly different. What is exactly expected here?

danielsjf committed:
Missing whitespace

MartinHjelmare reviewed (commented):
Looks good! Fix the lint issues and we can merge.

danielsjf committed:
Update co2signal.py

Fix pylint

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,112 @@
+"""
+Support for the CO2signal platform.
+For more details about this platform, please refer to the documentation
```
Add a blank line here.

MartinHjelmare commented on homeassistant/components/sensor/co2signal.py:
```diff
@@ -0,0 +1,112 @@
+Support for the CO2signal platform.
+For more details about this platform, please refer to the documentation
+"""
+
```
Remove this blank line.

danielsjf committed:
Update co2signal.py

Change blank lines

danielsjf committed:
Update co2signal.py

Initialise _data

MartinHjelmare reviewed (approved):
Great! Can be merged when build passes.

MartinHjelmare commented:
Next step is that I'll merge. :smile:

Thanks for the ping.
